### ISSUE TITLE
[SUGGESTION] Treat a keyword as an identifier

### ISSUE THREAD

**Issue by msadeqhe** [hsutter/cppfront#392]
### Preface

I suggest to use `"keyword"$` syntax to treat keywords as identifiers in Cpp2.

### Suggestion Detail

Some identifiers in Cpp2 are keywords in Cpp1 such as `and`, `or`, etc. These identifiers aren't valid identifiers in Cpp1. Therefore Cpp2 appends `cpp2_` prefix to these identifiers during code generation for Cpp1 as discussed in [this issue](https://github.com/hsutter/cppfront/issues/328). For example `and` in Cpp2 becomes `cpp2_and` in Cpp1.

On the other hand, some keywords in Cpp2 aren't keywords in Cpp1 such as `type`, `next`, etc. These keywords are valid identifiers in Cpp1. Therefore it's not possible to use such identifiers in Cpp2 when dealing with Cpp1 API:

```cpp
// ERROR! `type` is a keyword, whereas it's an identifier in Cpp1.
v0: = type.name();

// ERROR! `next` is a keyword, whereas it's an identifier in Cpp1.
while next < 10 next next++ {
    //: statements...
}
```

I suggest to use a syntax such as `"keyword"$` to access identifiers from Cpp1 API in which they are keywords in Cpp2:

```cpp
// ERROR! `type` is a keyword, whereas it's an identifier in Cpp1.
v0: = "type"$.name();

// ERROR! `next` is a keyword, whereas it's an identifier in Cpp1.
while "next"$ < 10 next "next"$++ {
    //: statements...
}
```

I think syntax `"keyword"$` is good enough because Cpp2 will have similar syntax for reflections and code generations as described in [this wiki](https://github.com/hsutter/cppfront/wiki/Design-note%3A-Capture).

### Your Questions

**Will your feature suggestion eliminate X% of security vulnerabilities of a given kind in current C++ code?**

No.

**Will your feature suggestion _automate or eliminate_ X% of current C++ guidance literature?**

No.

### Considered Alternatives

I considered to use syntax `@keyword` but it would be ambiguous with current syntax of meta-class functions.

**Comment by msadeqhe** [comment]
You're right :grimacing:, that's very nice but it doesn't work in every situation, for example:

```cpp
type: type = {}

main: () = {
    variable: type = 0;
}
```

**Comment by JohelEGP** [comment]
I prefer `cpp2_type`
over `"type"$`
to lower to Cpp1 as `type`.

**Comment by JohelEGP** [comment]
For `variable: type = ()`, you want `type` to be a Cpp1 identifier. In this case, it clashes with Cpp2 syntax itself. It can be solved with an indirection: `std::type_identity_t<type>`: <https://godbolt.org/z/TGoqsebrW>.

**Comment by JohelEGP** [comment]
This are the sets of Cpp2-specific keywords:
```C++
{i8,       i16,       i32,       i64,       u8,
 u16,      u32,       u64,       i8_fast,   i16_fast,
 i32_fast, i64_fast,  u8_fast,   u16_fast,  u32_fast,
 u64_fast, i8_least,  i16_least, i32_least, i64_least,
 u8_least, u16_least, u32_least, u64_least, inspect});
{next, copy, move, forward, pre, post, final,
 in, 
 as, is, type, 
 assert, throws,  
 implicit, out,   
 inout});            
```
I can't come up with another example like <https://github.com/hsutter/cppfront/issues/392#issuecomment-1519860597>'s.
A simpler alternative for `variable: type = ()`
is qualifying `type`: `variable: ::type = ()` <https://godbolt.org/z/W3v48se5K>.

**Comment by msadeqhe** [comment]
Thanks. It's a good idea to use helper type `std::type_identity_t<type>` or write the qualified name `::type`. But `::type` doesn't work for local types:

```cpp
main: () = {
    type: type = {}
    variable: ::type = ();
}
```

**Comment by MichaelCook** [comment]
Perhaps a backslash prefix would be a little more readable.

```
v0: = \type.name();
while \next < 10 next \next++ {
    //: statements...
}```

**Comment by JohelEGP** [comment]
Makes me think if raw string literals could just be backslash-escaped (see #302).

**Comment by msadeqhe** [comment]
@MichaelCook, I like the idea to use backslash syntax `\keyword`.

Escape sequences in string literals start with a backslash to change the meaning of a character, why not just use it to change the meaning of a keyword and treat it as an identifier?

@JohelEGP, Good idea but in my opinion, raw string literals break _the general capture syntax `(thing)$` everywhere in the language_. Also string literals without prefix or suffix will make it possible to have `operator''` and `operator""` or _Tagged Template Strings_ or any other versatile syntax in Cpp2.

**Comment by msadeqhe** [comment]
> @MichaelCook, I like the idea to use backslash syntax `\keyword`.

Because it only works on keywords, if Cpp2 has binary operator `\`, it wouldn't conflict with it:

```
// `type` is not a keyword.
x: \type = ();

// binary operator \
y: = var0 \ var1;
```

Also unary prefix and postfix operators `\` doesn't conflict with `\keyword` because they will be applied to identifiers.

**Comment by msadeqhe** [comment]
These are other use cases and reasons why it does matter in addition to C and Cpp1 language interop:

- When Cpp2 introduces a new keyword in the future, a simple find and replace from `keyword` to `\keyword` (or whatever else) helps programmers to migrate the source code of their programs and libraries.
- `\keyword` (or whatever else) is more readable than workarounds to treat a keyword as an identifier.
- It allows to use `keywords` for the name of interfaces and functions in APIs (other programming languages may use those names without any problem if they are not a keyword in their syntax):
```cpp
variable.\if(count < 10).\return();
```

**Comment by jcanizales** [comment]
> Therefore it's not possible to use such identifiers in Cpp2 when dealing with Cpp1 API

Could you give an example C++ API that would not be usable from Cpp2? Naming variables is not such a case.

**Comment by MichaelCook** [comment]
How's this:
```
struct myapi_t {
    int type;
};
```

**Comment by JohelEGP** [comment]
See https://github.com/hsutter/cppfront/issues/392#issuecomment-1519838852.

**Comment by msadeqhe** [comment]
Also all Cpp1 type traits have member type named `type`. For example:

```cpp
std::is_integral<T>::type
```

**Comment by msadeqhe** [comment]
Also `const` doesn't work as a type name, although both Cpp1 and Cpp2 have `const` keyword, so no problem for language interop:

```cpp
const: type = {}

main: () = {
    x: const = "";
}
```

But it works with qualified name. By the way it generates invalid Cpp1 code:

```cpp
const: type = {}

main: () = {
    x: ::const = "";
}
```

**Comment by msadeqhe** [comment]
> > Therefore it's not possible to use such identifiers in Cpp2 when dealing with Cpp1 API
> 
> I've actually found the Cpp2-specific keywords to be contextual, except for the `<cstdint>` shorthands (e.g., `i8`), and `inspect`. So this works: https://godbolt.org/z/vYPf6MG6h.
> 
> ```
> int next = 0; // Cpp1.
> f: () = {
>     while next < 10 next next++ {
>     //: statements...
>     }
>     while 10 < next next next++ {
>     //: statements...
>     }
> }
> main: () = {
>     :(is) = { [[assert: (is)is(is)]] }(:() = {});
> }
> ```
> 
> Including the `:(is) = { [[assert: (is)is(is)]] }(:() = {});`.

If only type names cannot be keywords, and if Cpp2 accepts keywords as type names with qualified names, I think I should close this issue:

```cpp
// OK.
variable: ::keyword = ///* definition *///

// OK.
function: ::keyword = ///* definition *///
```

Because it's a general rule and easy to understand/follow.

**Comment by msadeqhe** [comment]
To simlify the rule, maybe Cpp2 should require that every identifier to be qualified which has similar name to keywords:

```cpp
_::while: type = /* definition */
_::do: _::while = /* definition */

_::do._::type;

_::return(args);
```

So they must be always qualified either `::keyword` for global identifiers and `_::keyword` for local identifiers. Instead of `"keyword"$` or `@keyword` or `\keyword`, it's going to be `::keyword` or `_::keyword`.

These are alternative ways to achieve the same thing...

**Comment by msadeqhe** [comment]
> Instead of `"keyword"$` or `@"keyword"` or `\keyword`, it's going to be `namespace::keyword` or `::keyword` or `_::keyword`. Those are alternative ways to achieve the same thing...

Another option to consider is `keyword<>`. Types, functions and variables can be templates already, so `keyword<>` would mean a type, function or variable.

```cpp
if: (condition: bool, yes_value: int, no_value: int) -> int = {
    if condition {
        return yes_value;
    }
    else {
        return no_value;
    }
}

if: <T> (condition: bool, yes_value: T, no_value: T) -> T = {
    if condition {
        return yes_value;
    }
    else {
        return no_value;
    }
}

main: () = {
    x: = if<>(2 * 2 == 4, 1, 0);
    y: = if<bool>(2 * 2 == 4, true, false);
}
```

**Comment by msadeqhe** [comment]
> But keywords cannot be templates, so template arguments after keywords, mean they are not keywords.

Also it depends if Cpp2 would support `static_cast<X>(arg)`, `dynamic_cast<X>(arg)` and etc (template-style keywords) from Cpp1 or not.
